core/src/perf_libs.rs

```diff
@@ -0,0 +1,168 @@
+            use_non_default_stream: u8,
+        ) -> u32,
+    >,
+
+    pub chacha_cbc_encrypt_many_sample: Symbol<
+        'a,
+        unsafe extern "C" fn(
+            input: *const u8,
+            sha_state: *mut u8,
+            in_len: usize,
+            keys: *const u8,
+            ivec: *mut u8,
+            num_keys: u32,
+            samples: *const u64,
+            num_samples: u32,
+            starting_block: u64,
+            time_us: *mut f32,
+        ),
+    >,
+
+    pub chacha_init_sha_state: Symbol<'a, unsafe extern "C" fn(sha_state: *mut u8, num_keys: u32)>,
+    pub chacha_end_sha_state:
+        Symbol<'a, unsafe extern "C" fn(sha_state_in: *const u8, out: *mut u8, num_keys: u32)>,
+
+    pub poh_verify_many: Symbol<
+        'a,
+        unsafe extern "C" fn(
+            hashes: *mut u8,
+            num_hashes_arr: *const u64,
+            num_elems: usize,
+            use_non_default_stream: u8,
+        ) -> c_int,
+    >,
+
+    pub cuda_host_register:
+        Symbol<'a, unsafe extern "C" fn(ptr: *mut c_void, size: usize, flags: c_uint) -> c_int>,
+
+    pub cuda_host_unregister: Symbol<'a, unsafe extern "C" fn(ptr: *mut c_void) -> c_int>,
+}
+
+static mut API: Option<Container<Api>> = None;
+
+fn init(name: &OsStr) {
+    use std::sync::Once;
+    static INIT_HOOK: Once = Once::new();
+
+    info!("Loading {:?}", name);
+    unsafe {
+        INIT_HOOK.call_once(|| {
+            API = Some(Container::load(name).unwrap_or_else(|err| {
+                error!("Unable to load {:?}: {}", name, err);
+                std::process::exit(1);
+            }));
+        })
+    }
+}
+
+fn locate_perf_libs() -> Option<PathBuf> {
+    let exe = env::current_exe().expect("Unable to get executable path");
+    let perf_libs = exe.parent().unwrap().join("perf-libs");
+    if perf_libs.is_dir() {
+        info!("perf-libs found at {:?}", perf_libs);
+        return Some(perf_libs);
+    }
+    warn!("{:?} does not exist", perf_libs);
+    None
+}
+
+fn find_cuda_home(perf_libs_path: &Path) -> Option<PathBuf> {
+    // Search /usr/local for a `cuda-` directory that matches a perf-libs subdirectory
+    for entry in fs::read_dir(&perf_libs_path).unwrap() {
+        if let Ok(entry) = entry {
+            let path = entry.path();
+            if !path.is_dir() {
+                continue;
+            }
+            let dir_name = path.file_name().unwrap().to_str().unwrap_or("");
+            if !dir_name.starts_with("cuda-") {
+                continue;
+            }
+
+            let cuda_home: PathBuf = ["/", "usr", "local", dir_name].iter().collect();
+            if !cuda_home.is_dir() {
+                continue;
+            }
+
+            return Some(cuda_home);
+        }
+    }
+    None
+}
+
+pub fn init_cuda() {
+    if let Some(perf_libs_path) = locate_perf_libs() {
+        if let Some(cuda_home) = find_cuda_home(&perf_libs_path) {
+            info!("CUDA installation found at {:?}", cuda_home);
```

Comment:
This block encodes in Rust what previously was in `source target/perf-libs/env.sh`